### PR TITLE
REF: de-duplicate precision_from_unit attempt 2

### PR DESCRIPTION
**Pull request by jbrockmendel**
precision_from_unit is getting called in situations where it shouldn't be.

### CODE REVIEW

File: pandas/_libs/tslibs/conversion.pyx [hunk 1]
```diff
@@ -148,8 +145,8 @@ cdef int64_t cast_from_unit(
 
 
 cpdef inline (int64_t, int) precision_from_unit(
-        str unit,
-        NPY_DATETIMEUNIT out_reso=NPY_DATETIMEUNIT.NPY_FR_ns,
+    str unit,
+    NPY_DATETIMEUNIT out_reso=NPY_DATETIMEUNIT.NPY_FR_ns,
 ):
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/conversion.pyx:150]: won't this result in the same issue? `precision_from_unit` calls `get_conversion_factor`, which might error, but `precision_from_unit` isn't handling the exception

would one of the solutions from https://stackoverflow.com/questions/75544869/propagate-error-in-function-which-returns-tuple-of-int/75548002#75548002 work for that?

Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pyx:150]: The fix is that this moves the call to precision_from_unit down to after the Y/M cases are handled in cast_from_unit

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/conversion.pyx:150]: looks like it's still failing though
```
FAILED pandas/tests/indexes/period/test_indexing.py::TestGetIndexer::test_get_indexer2 - pytest.PytestUnraisableExceptionWarning: Exception ignored in: 'pandas._libs.tslibs.conversion.precision_from_unit'
```

Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pyx:150]: huh, not sure how i missed that locally.  will troubleshoot, thanks.

File: pandas/_libs/tslib.pyx [hunk 2]
```diff
@@ -275,8 +273,6 @@ def array_with_unit_to_datetime(
         )
         return result, tz
 
-    mult, _ = precision_from_unit(unit)
```
Comment (MarcoGorelli) [inline pandas/_libs/tslib.pyx:278]: 😄 is this unused in this function? I'd have hoped cython-lint would've caught this, I'll see what I can do there

File: pandas/_libs/tslibs/conversion.pyx [hunk 3]
```diff
@@ -104,12 +105,6 @@ cdef int64_t cast_from_unit(
         int64_t m
         int p
 
-    m, p = precision_from_unit(unit, out_reso)
-
-    # just give me the unit back
-    if ts is None:
-        return m
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/conversion.pyx:111]: is it because this was uncovered anyway?

File: pandas/_libs/tslibs/conversion.pyx [hunk 4]
```diff
@@ -166,34 +163,21 @@ cpdef inline (int64_t, int) precision_from_unit(
...
-        m = multiplier
-    elif reso == NPY_DATETIMEUNIT.NPY_FR_ms:
-        m = multiplier // 1_000
-    elif reso == NPY_DATETIMEUNIT.NPY_FR_us:
-        m = multiplier // 1_000_000
-    elif reso == NPY_DATETIMEUNIT.NPY_FR_ns or reso == NPY_DATETIMEUNIT.NPY_FR_GENERIC:
-        m = multiplier // 1_000_000_000
     else:
-        raise ValueError(f"cannot cast unit {unit}")
+        m = get_conversion_factor(reso, out_reso)
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/conversion.pyx:179]: if `get_conversion_factor` raises, then `precision_from_unit` wouldn't propagate the error - perhaps a note could be made about this in the docstring, alongside the current note?

**Review by MarcoGorelli — APPROVED** [review]
Looks good - left a comment about the error from `precision_from_unit` not propagating through `get_conversion_factor`, but feel free to ignore if you disagree, not a blocker

File: pandas/_libs/tslib.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx:278]: correct

File: pandas/_libs/tslibs/conversion.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pyx:111]: correct

File: pandas/_libs/tslibs/conversion.pyx (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pyx:179]: comment sounds good

Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pyx:179]: to keep from clogging the CI, mind if i add this comment in my next CLN: assorted branch?

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/conversion.pyx:179]: sounds good!

**Pull request merged**